californium-core/src/main/java/org/eclipse/californium/core/network/CoapEndpoint.java

```diff
@@ -465,8 +519,28 @@ public void sendEmptyMessage(final Exchange exchange, final EmptyMessage message
 	 *  @throws NullPointerException if the given deliverer is {@code null}
 	 */
 	@Override
-	public void setMessageDeliverer(MessageDeliverer deliverer) {
-		coapstack.setDeliverer(deliverer);
+	public void setMessageDeliverer(final MessageDeliverer deliverer) {
+		if (deliverer == null) {
+			throw new NullPointerException("Message deliverer must not be null");
+		} else {
+			coapstack.setDeliverer(new MessageDeliverer() {
+
+				@Override
+				public void deliverResponse(final Exchange exchange, final Response response) {
+					deliverer.deliverResponse(exchange, response);
+					if (response.isNotification()) {
```

Comment:
Why send it to both the local Response handler and the registered NotificationListener?